### PR TITLE
Math Magicians: Test

### PR DESCRIPTION
**Pull request by mulugithub**
### Hello CR 🤝 

_In this PR, we implement the following features:_

**General requirements**

- [ ]  [No linter errors](https://github.com/microverseinc/linters-config).
- [ ] Used correct [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
- [ ] Documented the work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).
- [ ] Followed the list of [best practices for JavaScript](https://github.com/microverseinc/curriculum-html-css/blob/main/articles/javascript_best_practices.md).

**Project Requirements**

- [ ] Set up React Testing Library.
- [ ] Unit tests for the files operate.js and calculate.js using Jest.
- [ ] Create unit tests for all React components:
- [ ] Used Jest snapshots to test the components.
- [ ] Used React Testing Library to simulate user interaction.
- [ ] Run full tests suite using Jest. All tests passed.
- [ ]

### CODE REVIEW

File: src/test/buttons.test.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+describe('Buttons component', () => {
+  test('Test Calculator Buttons', () => {
+    const tree = render(
+      <BrowserRouter>
+        <Buttons onButtonClick={jest.fn()} />
+      </BrowserRouter>,
+    );
+    expect(tree).toMatchSnapshot();
+  });
+});
```
Comment (KanzaTahreem) [inline src/test/buttons.test.js:15]: - [ ] Great job on incorporating snapshot tests! 👏 However, I've observed that you are currently lacking tests that simulate user interactions using the React Testing Library as mentioned in the [project requirements](https://github.com/microverseinc/curriculum-react-redux/blob/main/math-magicians/project_tests.md) ⛏🤓 Snapshot tests are excellent for catching unexpected UI changes, yet they don't provide insights into how users actually interact with your application. By including tests that utilize the React Testing Library, you can accurately simulate user actions like button clicks, input field entries, and more. For example


```jsx
it('handles operator button clicks correctly', () => {
  const mockButtonClick = jest.fn();
  const { getByText } = render(<Buttons onButtonClick={mockButtonClick} />);
  fireEvent.click(getByText('÷'));
  fireEvent.click(getByText('x'));
  waitFor(() => expect(mockButtonClick).toHaveBeenCalledTimes(2));
  waitFor(() => expect(mockButtonClick).toHaveBeenCalledWith('÷'));
});
```



  image

You can learn more about the react testing library [here](https://testing-library.com/docs/react-testing-library/example-intro) ℹ

**_I noticed that `<Home />` is tested well. Please make sure to use React Testing Library to test the remaining components where it hasn't been implemented._**

File: src/test/calculator.test.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+describe('Calculator component', () => {
+  test('renders correctly', () => {
+    const tree = render(
+      <BrowserRouter>
+        <Calculator />
+      </BrowserRouter>,
+    );
+    expect(tree).toMatchSnapshot();
+  });
+});
```
Comment (KanzaTahreem) [inline src/test/calculator.test.js:15]: - [ ] You can consider the following 👇 Please note that this is just an example. You are not obliged to use the provided example; however, please ensure that the intended purpose (mentioned above) is met.

```jsx
it('should perform addition when + button is clicked', () => {
  const { getByText } = render(<Calculator />);
  fireEvent.click(getByText('1'));
  fireEvent.click(getByText('+'));
  fireEvent.click(getByText('2'));
  fireEvent.click(getByText('='));
  waitFor(() => expect(getByText('3')).toBeInTheDocument());
});
```

**_Please make sure to use React Testing Library to test the remaining components where it hasn't been implemented._**

File: src/test/operate.test.js
Comment (KanzaTahreem) [inline src/test/operate.test.js:1]: _Well done for covering all the functionalities & scenarios 🎉 Keep rocking 🚀_

**Review by KanzaTahreem — CHANGES_REQUESTED** [review]
# Changes Required 🔁

Hi @mulugithub & @the-faizmohammad 

Good job so far 👏! I really appreciate the effort you put in!  **_However, there are few this that still need to be fixed. But you are almost there!_** 💪 🏁

so close


## Highlights 🎯

There are some excellent points that you have made perfectly. Keep up the great work! 👏
- ✅ Implemented tests are passing
- ✅ Use Jest snapshots to test the components
- ✅ Tested `operate.js` and `calculate.js` using Jest

## Required Changes 🛑

_Please check the comments under the review._

## Optional suggestions ❕

N/A

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👨‍💻💻

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Alusp — APPROVED** [review]
Hi @mulugithub and @the-faizmohammad , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

congrats

✅ Implemented tests are passing
✅ Use Jest snapshots to test the components
✅ Tested operate.js and calculate.js using Jest

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @Alusp in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**